storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+		}
+		r := bytes.NewReader(b)
+		logrus.Debugf("exporting opaque data as blob %q", digest)
+		return ioutil.NopCloser(r), r.Size(), nil
+	}
+	logrus.Debugf("exporting filesystem layer as blob %q", digest)
+	return s.diffLayer(s.Layers[digest], true)
+}
+
+func (s *storageImage) diffLayer(layerID string, computeSize bool) (rc io.ReadCloser, n int64, err error) {
+	layer, err := s.store.GetLayer(layerID)
+	if err != nil {
+		return nil, -1, err
+	}
+	layerMeta := storageLayerMetadata{
+		ExpectedSize: -1,
+	}
+	if layer.Metadata != "" {
+		if err := json.Unmarshal([]byte(layer.Metadata), &layerMeta); err != nil {
+			logrus.Errorf("error decoding metadata for layer %q: %v", layerID, err)
+			return nil, -1, err
+		}
+	}
+	if computeSize {
+		if layerMeta.ExpectedSize == -1 {
+			n, err = s.store.DiffSize("", layer.ID)
+			if err != nil {
+				return nil, -1, err
+			}
+		} else {
+			n = layerMeta.ExpectedSize
+		}
+	} else {
+		n = -1
+	}
+	diff, err := s.store.Diff("", layer.ID)
+	if err != nil {
+		return nil, -1, err
+	}
+	return diff, n, nil
+}
+
+func (s *storageImage) GetManifest() (manifest []byte, MIMEType string, err error) {
+	manifest, err = s.store.GetImageBigData(s.ID, "manifest")
+	return manifest, "", err
+}
+
+func (s *storageImage) GetTargetManifest(digest string) (manifest []byte, MIMEType string, err error) {
+	requested, err := ddigest.ParseDigest(digest)
+	if err != nil {
+		return nil, "", err
+	}
+	manifest, err = s.store.GetImageBigData(s.ID, "manifest")
+	manifestDigest := requested.Algorithm().FromBytes(manifest)
+	if manifestDigest != requested {
+		return nil, "", ErrManifestDigestMismatch
+	}
+	return manifest, "", err
+}
+
+func (s *storageImage) GetSignatures() (signatures [][]byte, err error) {
+	var offset int
+	if err := s.loadMetadata(); err != nil {
+		logrus.Errorf("error decoding metadata for image: %v", err)
+		return nil, err
+	}
+	signature, err := s.store.GetImageBigData(s.ID, "signatures")
+	if err != nil {
+		return nil, err
+	}
+	sigslice := [][]byte{}
+	for _, length := range s.SignatureSizes {
+		sigslice = append(sigslice, signature[offset:offset+length])
+		offset += length
+	}
+	if offset != len(signature) {
+		return nil, fmt.Errorf("signatures data contained %d extra bytes", len(signatures)-offset)
+	}
+	return sigslice, nil
+}
+
+func (s *storageImage) DeleteImage() error {
+	if s.ID != "" {
+		if _, err := s.store.DeleteImage(s.ID, true); err != nil {
+			return err
+		}
+		s.ID = ""
+	}
+	return nil
+}
+
+func (s *storageImage) IsMultiImage() bool {
+	return false
+}
+
+func (s *storageImage) Manifest() (manifest []byte, MIMEType string, err error) {
+	return s.GetManifest()
+}
+
+func (s *storageImage) Signatures() (signatures [][]byte, err error) {
+	return s.GetSignatures()
+}
+
+func (s *storageImage) ShouldCompressLayers() bool {
+	// The storage layer automatically decompresses layers as part of
+	// applying them, but data blobs might as well be compressed on disk.
+	return true
```

Comment:
Does this mean that the storage layer _both_ maintains a FS with the layers applied, _and_ the original blobs?